cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTest.java

```diff
@@ -284,9 +284,133 @@ public void apply() throws Exception {
     });
   }
 
+
+  @Test
+  public void testPartitionConsumingWithFilterAndLimit() throws Exception {
+    final PartitionedFileSet dataset = dsFrameworkUtil.getInstance(pfsInstance);
+    final TransactionAware txAwareDataset = (TransactionAware) dataset;
+
+    final Set<PartitionKey> partitionKeys1 = Sets.newHashSet();
+    for (int i = 0; i < 10; i++) {
+      partitionKeys1.add(generateUniqueKey());
+    }
+
+    final Set<PartitionKey> partitionKeys2 = Sets.newHashSet();
+    for (int i = 0; i < 15; i++) {
+      partitionKeys2.add(generateUniqueKey());
+    }
+
+    final PartitionConsumer partitionConsumer = new PartitionConsumer(dataset);
+    // add each of partitionKeys1 in separate transaction, so limit can be applied at arbitrary values
+    // (consumption only happens at transaction borders)
+    for (final PartitionKey partitionKey : partitionKeys1) {
+      dsFrameworkUtil.newInMemoryTransactionExecutor(txAwareDataset).execute(new TransactionExecutor.Subroutine() {
+        @Override
+        public void apply() throws Exception {
+            dataset.getPartitionOutput(partitionKey).addPartition();
+        }
+      });
+    }
+
+    dsFrameworkUtil.newInMemoryTransactionExecutor(txAwareDataset).execute(new TransactionExecutor.Subroutine() {
+      @Override
+      public void apply() throws Exception {
+        // Initial consumption results in the partitions corresponding to partitionKeys1 to be consumed because only
+        // those partitions are added to the dataset at this point
+        List<Partition> consumedPartitions = Lists.newArrayList();
+
+        // with limit = 1, the returned iterator is only size 1, even though there are more unconsumed partitions
+        Iterators.addAll(consumedPartitions, partitionConsumer.consumePartitions(PartitionFilter.ALWAYS_MATCH, 1));
+        Assert.assertEquals(1, consumedPartitions.size());
+
+        // ask for 5 more
+        Iterators.addAll(consumedPartitions, partitionConsumer.consumePartitions(PartitionFilter.ALWAYS_MATCH, 5));
+        Assert.assertEquals(6, consumedPartitions.size());
+
+        // ask for 5 more, but there are only 4 more unconsumed partitions (size of partitionKeys1 is 10).
+        Iterators.addAll(consumedPartitions, partitionConsumer.consumePartitions(PartitionFilter.ALWAYS_MATCH, 5));
+        Assert.assertEquals(10, consumedPartitions.size());
+
+        Set<PartitionKey> retrievedKeys = Sets.newHashSet();
+        for (Partition consumedPartition : consumedPartitions) {
+          retrievedKeys.add(consumedPartition.getPartitionKey());
+        }
+        Assert.assertEquals(partitionKeys1, retrievedKeys);
+      }
+    });
+
+    dsFrameworkUtil.newInMemoryTransactionExecutor(txAwareDataset).execute(new TransactionExecutor.Subroutine() {
+      @Override
+      public void apply() throws Exception {
+        for (PartitionKey partitionKey : partitionKeys2) {
+          dataset.getPartitionOutput(partitionKey).addPartition();
+        }
```

Comment:
like! Nice that you test both cases, the one where a transaction adds a single partition, and also where it adds multiple.